mypyc/emitmodule.py

```diff
@@ -63,57 +74,122 @@ def __init__(self, declaration: HeaderDeclaration, mark: bool) -> None:
         self.mark = False
 
 
+class MypycPlugin(Plugin):
+    """Plugin for making mypyc interoperate properly with mypy incremental mode.
+
+    Basically the point of this plugin is to force mypy to recheck things
+    based on the demands of mypyc in a couple situations:
+      * Any modules in the same group must be compiled together, so we
+        tell mypy that modules depend on all their groupmates.
+      * If the IR metadata is missing or stale or any of the generated
+        C source files associated missing or stale, then we need to
+        recompile the module so we mark it as stale.
+    """
+
+    def __init__(self, options: Options, groups: Groups) -> None:
+        super().__init__(options)
+        self.group_map = {}  # type: Dict[str, Tuple[Optional[str], List[str]]]
+        for sources, name in groups:
+            modules = sorted(source.module for source in sources)
+            for id in modules:
+                self.group_map[id] = (name, modules)
+
+        self.metastore = create_metastore(options)
+
+    def report_config_data(
+            self, ctx: ReportConfigContext) -> Optional[Tuple[Optional[str], List[str]]]:
+        # The config data we report is the group map entry for the module.
+        # If the data is being used to check validity, we do additional checks
+        # that the IR cache exists and matches the metadata cache and all
+        # output source files exist and are up to date.
+
+        id, path, is_check = ctx.id, ctx.path, ctx.is_check
+
+        if id not in self.group_map:
+            return None
+
+        # If we aren't doing validity checks, just return the cache data
+        if not is_check:
+            return self.group_map[id]
+
+        # Load the metadata and IR cache
+        meta_path, _, _ = get_cache_names(id, path, self.options)
+        ir_path = get_ir_cache_name(id, path, self.options)
+        try:
+            meta_json = self.metastore.read(meta_path)
+            ir_json = self.metastore.read(ir_path)
+        except FileNotFoundError:
+            return None
+
+        ir_data = json.loads(ir_json)
+
+        # Check that the IR cache matches the metadata cache
+        if compute_hash(meta_json) != ir_data['meta_hash']:
+            return None
+
+        # Check that all of the source files are present and as expected
```

Comment:
Maybe elaborate a bit about when they may not be present or as expected (i.e. is this a normal condition or has something been corrupted).

Reply:
Will do. Main situation is if the user deleted `build/` without deleting `.mypy_cache`, which is sort of corruption but also reasonable to expect that it will happen.